File: kotlin/src/main/kotlin/knn.kt
import org.locationtech.jts.geom.Geometry
import org.locationtech.jts.index.strtree.GeometryItemDistance
import org.locationtech.jts.index.strtree.STRtree
import org.locationtech.jts.io.WKTReader
import java.io.File
import java.sql.DriverManager
import java.sql.SQLException


class DbManager(
    user: String,
    pass: String,
    host: String,
    db: String
) {
    private val connectionUrl = "jdbc:postgresql://$host/$db?user=$user&password=$pass"

    fun getTable(sql: String): Map<String, Geometry> {
        val geom = mutableMapOf<String, Geometry>()
        val reader = WKTReader()

        try {
            DriverManager.getConnection(this.connectionUrl).use {
                connection -> connection.createStatement().use {
                    statement -> statement.executeQuery(sql).use { resultSet ->
                        while (resultSet.next()) {
                            geom[resultSet.getString("id")] = reader.read(resultSet.getString("geom"))
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            e.printStackTrace()
        }
        return geom.toMap()
    }
}
fun nearestNeighbour(geoma: Map<String, Geometry>, geomb: Map<String, Geometry>): Map<String, Pair<String?, Double>> {
    //for each geometry a get entry of b with the lowest distance, then compute dist to save map
    val dist: Map<String, Pair<String?, Double>> = geoma.map {
            a ->
                val knn = geomb.minWithOrNull(compareBy({ b -> a.value.distance(b.value)},{ b -> b.key}))
                a.key to (knn?.key to a.value.distance(knn?.value))
    }.toMap()
    return dist
}

fun nearestNeighbour2(geoma: Map<String, Geometry>, geomb: Map<String, Geometry>): Map<String, Pair<String?, Double>> {
    val t = STRtree()
    geomb.forEach {
        t.insert(it.value.envelopeInternal, it.value)
    }
    t.build()
    val geomb2 = geomb.map{ x -> x.value to x.key}.toMap()

    val dist = geoma.map {
            a ->
                val knnGeom = t.nearestNeighbour(a.value.envelopeInternal, a.value, GeometryItemDistance(), 100).toList() as List<Geometry>
                val knn = knnGeom.associate { y -> geomb2[y] to a.value.distance(y) }
                    .minWithOrNull(compareBy({ b -> b.value },{ b -> b.key}))
                a.key to (knn?.key to knn?.value)
    }.toMap() as Map<String, Pair<String?, Double>>

    return dist
}

fun saveCsv(table: Map<String, Pair<String?, Double>>, name: String) {
    val writer = File("$name").bufferedWriter()
    writer.write("""origin,destination,distance""")
    writer.newLine()
    table.forEach {
        writer.write("""${it.key}, "${it.value.first}", ${it.value.second}""")
        writer.newLine()
    }
    writer.flush()
}

fun main() {
    println("db user")
    val user = readln()
    println("bd password")
    val pass = readln()

    val db = DbManager(user, pass, "localhost", "gis")
    val sql1 = """SELECT "UPRN" id, ST_AsText(geom) geom FROM os.open_uprn_white_horse"""
    val sql2 = """SELECT "postcode" id, ST_AsText(geom) geom FROM os.code_point_open_white_horse"""

    val uprn = db.getTable(sql1)
    val codepoint = db.getTable(sql2)

    val startTime = System.currentTimeMillis()
    val out1 = nearestNeighbour(uprn, codepoint) //22sec
    val endTime = System.currentTimeMillis()
    saveCsv(out1, "kotlin_all_vs_all.csv")

    println(startTime - endTime)
    val startTime2 = System.currentTimeMillis()
    val out2 = nearestNeighbour2(uprn, codepoint) //3.6sec
    val endTime2 = System.currentTimeMillis()
    saveCsv(out2, "kotlin_tree.csv")

    println(startTime2 - endTime2)
}

File: scala/src/main/scala/knn.scala
import org.locationtech.jts.geom.Geometry
import org.locationtech.jts.index.strtree.{GeometryItemDistance, STRtree}
import org.locationtech.jts.io.WKTReader

import java.io.{BufferedWriter, File, FileWriter}
import java.sql.{DriverManager, SQLException}
import java.util.Locale
import scala.collection.immutable.HashMap
class DbManager(
                 user: String,
                 pass: String,
                 host: String,
                 port: String,
                 db: String
               ) :
  val connectionUrl = s"jdbc:postgresql://$host:$port/$db?user=$user&password=$pass"

  def getTable(sql: String) =
    val geom = scala.collection.mutable.HashMap.empty[String, Geometry]
    val reader = WKTReader()

    try
      val connection = DriverManager.getConnection(connectionUrl)
      val statement = connection.createStatement()
      val resultSet = statement.executeQuery(sql)

      while resultSet.next do
              geom += resultSet.getString("id") -> reader.read(resultSet.getString("geom"))
      connection.close()
    catch
        case e: SQLException => e.printStackTrace

    geom.toMap

def nearestNeighbour(geoma: Map[String, Geometry], geomb: Map[String, Geometry]) =
  //for each geometry a get entry of b with the lowest distance, then compute dist to save map
  geoma
    .map(
      x =>
        val knn = geomb.minBy(b => (x._2.distance(b._2), b._1))
        (x._1 , knn._1, x._2.distance(knn._2))
    )
    .toList

def nearestNeighbour2(geoma: Map[String, Geometry], geomb: Map[String, Geometry]) =
  val t = STRtree()
  geomb.foreach( x =>
    t.insert( x._2.getEnvelopeInternal, x._2)
  )
  t.build()
  val geomb2 = geomb.map(_.swap)

  geoma.map (
    x =>
      val knnGeom = t.nearestNeighbour(x._2.getEnvelopeInternal, x._2, GeometryItemDistance(), 100).toList.asInstanceOf[List[Geometry]]
      val knn = knnGeom.map(y => (geomb2(y), x._2.distance(y))).reduceLeft { (a, b) =>
        if math.abs(a._2 - b._2) < 1e-9 then
          if a._1 < b._1 then a else b
        else if a._2 < b._2 then a else b
      }

      (x._1, knn._1, knn._2)
    )
    .toList

def saveCsv(table: List[(String, String, Double)], name: String) =
  val file = File(s"$name")
  val writer = BufferedWriter(FileWriter(file))
  writer.write("""origin,destination,distance""")
  writer.newLine()
  table.foreach(
    x =>
      writer.write(s"${x(0)},\"${x(1)}\",${String.format(Locale.US, "%f", x(2))}")
      writer.newLine()
  )
  writer.flush()

@main def main(args: String*) =
  val mode = args.headOption.getOrElse("both")

  val user   = Option(System.getenv("DB_USER")).getOrElse("postgres")
  val pass   = Option(System.getenv("DB_PASSWORD")).getOrElse("")
  val host   = Option(System.getenv("DB_HOST")).getOrElse("localhost")
  val port   = Option(System.getenv("DB_PORT")).getOrElse("5432")
  val dbName = Option(System.getenv("DB_NAME")).getOrElse("gis")

  val db   = DbManager(user, pass, host, port, dbName)
  val uprnTable      = Option(System.getenv("UPRN_TABLE")).getOrElse("os.open_uprn_white_horse")
  val codepointTable = Option(System.getenv("CODEPOINT_TABLE")).getOrElse("os.code_point_open_white_horse")
  val sql1 = s"SELECT uprn::text id, ST_AsText(geom) geom FROM $uprnTable"
  val sql2 = s"SELECT postcode id, ST_AsText(geom) geom FROM $codepointTable"

  val uprn      = db.getTable(sql1)
  val codepoint = db.getTable(sql2)

  val timingsWriter = BufferedWriter(FileWriter(File("timings.csv")))
  timingsWriter.write("test,elapsed_s")
  timingsWriter.newLine()

  if mode == "brute" || mode == "both" then
    val startTime = System.currentTimeMillis()
    val out1 = nearestNeighbour(uprn, codepoint)
    val endTime = System.currentTimeMillis()
    saveCsv(out1, "scala_all_vs_all.csv")
    timingsWriter.write(s"Scala all vs all,${(endTime - startTime) / 1000.0}")
    timingsWriter.newLine()

  if mode == "tree" || mode == "both" then
    val startTime = System.currentTimeMillis()
    val out2 = nearestNeighbour2(uprn, codepoint)
    val endTime = System.currentTimeMillis()
    saveCsv(out2, "scala_tree.csv")
    timingsWriter.write(s"Scala strtree,${(endTime - startTime) / 1000.0}")
    timingsWriter.newLine()

  timingsWriter.flush()